exercises/practice/parallel-letter-frequency/.meta/proof.ci.js

```diff
@@ -0,0 +1,19 @@
+export const parallelLetterFrequency = (texts) => {
+  let result = {};
+  let formatedTexts = texts
+    .map((x) => x.toLowerCase().match(/\p{Letter}+/gu) ?? [])
+    .flat();
+  Promise.all(formatedTexts.map((t) => processSingleText(t, result)));
+  return result;
+};
+
+const processSingleText = (text, result) => {
+  let res = [...text].reduce((acc, cur) => {
+    acc[cur] = (acc[cur] || 0) + 1;
+    return acc;
+  }, result);
+
+  return new Promise((resolve) => {
+    resolve(res);
+  });
+};
```

Comment:
I'm not a JavaScript maintainer, but to me this looks like the processing isn't actually done in parallel but calculated sequentially and then wrapped in a promise. I think if you move the processing code within the callback in the `Promise` constructor, then it would do parallel processing.

Reply:
There is no way to process things in parallel in JavaScript because JavaScript is single threaded. You _can_ however process things concurrently. Not that the code will achieve it, because all this does is spawn micro tasks on the stack and the evaluate them in order, it would be "close enough".

Moving the processing to the `Promise` constructor doesn't change anything. It just defers the processing to inside the microtask.

Reply:
Yeah, I'm aware of the single thread limitation, I wrote about it a bit in the Discord and ultimately decided to cheat a little with `Promise.all()`. 
According to MDN however, you can use [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) for multi thread processing, as each worker is spawned in a separate thread. Something like this maybe (I haven't tested this code, yet)
```js
// worker.js
onmessage = function (event) {
  let res = [...event.data].reduce((acc, cur) => {
    acc[cur] = (acc[cur] || 0) + 1;
    return acc;
  }, {});
  postMessage(res);
};
```
```js
// main.js
function workerDemo(text) { 
    const worker = new Worker('worker.js');
    const result = {}
    worker.postMessage(text);
    worker.onmessage = function (event) {
        result = event.data;
    };
    return result;
};
```

In theory, if you split the initial array in two, or however many, and then spawn a separate worker for each part, they should execute each in parallel. But my problems with `Web Workers` are that:
1. I don't know how to test them properly
2. I don't want to write tests that make it impossible to pass if you don't use `Web Workers`
3. I think that the learning curve is too big so it'll be a bit of a hurdle for most people without some further reading (speaking from my experience from mentoring, most requests i get are for simpler exercises so...)

I could (probably would) write up some explanation, about `Web Workers` but yeah, your input is greatly appreciated.

Reply:
@SleeplessByte does it still make sense to implement this exercise then?

Reply:
@ErikSchierboom with the additions from @Cool-Katt I like it. It's a good primer that JS _is_ single threaded by default. I think it's fine like this.